### DISCUSSION TITLE
How does `get_user` make sure that the session is still valid (not expired)?

### DISCUSSION THREAD

**Discussion by imran-iq** [django/channels#2077] [Help!] [answered]
In the docs it is mentioned that we should periodically check with `get_user(scope)` in long lived sessions to make sure that the session is still valid. 

From: https://asgi.readthedocs.io/en/latest/specs/main.html
`A connection scope, which represents a protocol connection to a user and survives until the connection closes.`

So my (naive) understanding is that a connection is started, and the session is put into `self.scope` of the consumer.
Inside the consumer we call `get_user(self.scope)`: https://github.com/django/channels/blob/b6dc8c127d7bda3f5e5ae205332b1388818540c5/channels/auth.py#L20-L55

which does the following: 
1. check there is a session object in the scope
2. check that the users password has not changed with the auth hash key

but it does not check that the session has expired, i.e. in the following case:
1. a session is created that expires in lets say 1 hour
2. a user uses the session cookie to connect to the websocket and the session is set in the scope
3. 2 hours pass
4. calling get_user(scope) will still return the authenticated user even though the session has expired since the expiration time is not checked
5. The user remains authenticated in the long lived connection despite having an expired session

So until the user closes their websocket connection, `get_user` will continue to think the user is authenticated

Or is there some other mechanism that ensures the auth hash changes once a session is expired?

**Comment by carltongibson** [comment, marked as answer]
If you're dealing with short-lived sessions you can check the session's [expire_date](https://docs.djangoproject.com/en/5.0/topics/http/sessions/#django.contrib.sessions.base_session.AbstractBaseSession.expire_date), or you can refresh the session in the `scope` (similarly to refetching the user). 

(I suspect periodically calling `session.save()` is going to be the simplest way forward, but it depends on your exact use-case 🤔)

**Reply by imran-iq** [reply]
Yeah I have ended up checking `expire_date` along with refetching the user just in case :+1:
